File: company/Facebook/arrays/Twosum.py
class Solution(object):
    def twoSum( nums, target):
        dict = {}
        for i in range(len(nums)):
            if dict.get(target-nums[i]) is not None:
                return [dict.get(target-nums[i]), i]
            else:
                dict[nums[i]]=i

    if __name__ == '__main__':
        num = [2,7,6,5]
        
        x = twoSum( num, 9)
        print(x)


# same logic 0(n)

File: company/Facebook/arrays/Twosum.java
import java.util.HashMap;
import java.util.Map;

// use hashmap--save the num[i] with its iindex in hashmap
// for evry loop check if compliment is present , if it present return the current number index and the get the compliment s value from hashmap
// k,v ==> num[i], index
// Time 0(n)
//Worst case fo reach element we store, we get the compliment 0(1) n times
// for hashing its only 0(1)

class Twosum {
    public static int[] twoSum(int[] nums, int target) {
        
        Map< Integer, Integer> hashmap = new HashMap<>();
        for (int i=0; i< nums.length; i++){
            int compliment = target-nums[i];
            
            if(hashmap.containsKey(compliment)){
                return new int[] {hashmap.get(compliment),  i};
            }
            
            hashmap.put(nums[i], i);
        }
        throw new IllegalArgumentException("No Such solution is present");
        
    }
        
    public static void main(String[] args) {
        
       int num [] = {2, 7, 8, 6};
    
  
        System.out.println(twoSum(num, 9));
     }
}


